pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -624,6 +698,42 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, ru
 	return
 }
 
+func (m *kubeGenericRuntimeManager) preStopSidecars(pod *v1.Pod, containers []*kubecontainer.Container) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(containers))
+
+	for _, container := range containers {
+		go func(container *kubecontainer.Container) {
+			defer wg.Done()
+			gracePeriod := int64(minimumGracePeriodInSeconds)
+			switch {
+			case pod.DeletionGracePeriodSeconds != nil:
+				gracePeriod = *pod.DeletionGracePeriodSeconds
+			case pod.Spec.TerminationGracePeriodSeconds != nil:
+				gracePeriod = *pod.Spec.TerminationGracePeriodSeconds
+			}
+			var containerSpec *v1.Container
+			if pod != nil {
+				if containerSpec = kubecontainer.GetContainerSpec(pod, container.Name); containerSpec == nil {
```

Comment:
I think there is a case where, even if the pod is not nil, the containerSpec we get from here may be nil and we may need to restore the container spec. We needed to add a check for this here: https://github.com/lyft/kubernetes/pull/5/files. Might be something we need here as well? Not sure if you've seen nil podSpecs in your testing, but it should be easy to reproduce if you restart the kubelet.

Reply:
I'm just working on implementing this, I'm gonna add some testcases for when the Pod is nil to ensure we don't get any panics.